src/hooks/useNotificationSound.ts

```diff
@@ -0,0 +1,70 @@
+import { useEffect, useState } from "react";
+import useSound from "use-sound";
+
+import notificationSound from "assets/sounds/newmessage.m4a";
+
+import { PrivateChatMessage, ChatTypes } from "types/chat";
+
+import { useChatSidebarControls } from "hooks/chatSidebar";
+
+import { useUser } from "./useUser";
+import { WithId } from "utils/id";
+
+export const useNotificationSound = (
+  privateChatMessages?: WithId<PrivateChatMessage>[]
+) => {
+  const { user } = useUser();
+  const userId = user?.uid;
+
+  const { chatSettings } = useChatSidebarControls();
+
+  const [play] = useSound(notificationSound, {
+    // `interrupt` ensures that if the sound starts again before it's
+    // ended, it will truncate it. Otherwise, the sound can overlap.
+    interrupt: true,
+  });
+
+  const [privateChatMessagesState, setPrivateChatMessagesState] = useState(
+    privateChatMessages
+  );
```

Comment:
I'm guessing that we're using `privateChatMessagesState` to hold the previous value of `privateChatMessages` here, so that we can then compare the messages later, to determine if we need to play the sound?

If so, since we don't need to force re-renders when we update the data, it might be better to use `useRef` here rather than `useState`:

- https://reactjs.org/docs/hooks-reference.html#useref

That's that basic pattern that utility hooks such as the following (which we don't currently have in our dependencies) use:

- https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state
  - based on this pattern
- https://github.com/streamich/react-use/blob/master/docs/usePrevious.md
- https://github.com/streamich/react-use/blob/master/docs/usePreviousDistinct.md

Reply:
I'm not sure we want to keep an entire copy of all of the private chat messages in memory just to determine if there is a new one though. Could we just store the last updated time/similar, and then filter/process only messages received after that instead?

Reply:
`PrivateChatMessage` extends from `BaseChatMessage` which has a `ts_utc` (`firebase.firestore.Timestamp`) field:

- https://github.com/sparkletown/sparkle/blob/staging/src/types/chat.ts#L4-L14

The first time this hook is run, we can sort `privateChatMessages` by `ts_utc` (see `chatSort` in `src/utils/chat.ts`), take the newest, and store that message in the `useRef`.

Then, the next time the hook is run, we can check that we have a value stored in the `useRef`, we can filter `privateChatMessages` to exclude anything older than that stored message.

If there are no newer messages, we can probably bail out at this point.

If there are newer messages, we want to:

- sort them by `ts_utc` (see `chatSort` in `src/utils/chat.ts`), get the newest timestamp, and then store that in our `useRef` (updating the `useRef` might be the last thing we do)
- find at least one value ([`.some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)) that matches our rules for 'should play a sound'
  - eg. message isn't read, message isn't sent by us, user isn't in chat with the person the message is from
  - If there is a matching value, play the sound
  
---

This should probably be in a `useEffect` or similar, but this is a rough snippet of how it will probably end up looking:

```typescript
const shouldPlaySound = privateChatMessages.some((message) => !message.isRead && ![userId, currentChatUserId].includes(message.from))

if (shouldPlaySound) {
  play()
}
```